src/main/java/net/imagej/ops/filter/vesselness/DefaultFrangi.java

```diff
@@ -0,0 +1,282 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package net.imagej.ops.filter.vesselness;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+
+import Jama.EigenvalueDecomposition;
+import Jama.Matrix;
+import net.imagej.ops.Ops;
+import net.imagej.ops.special.computer.AbstractUnaryComputerOp;
+import net.imglib2.Cursor;
+import net.imglib2.RandomAccess;
+import net.imglib2.RandomAccessibleInterval;
+import net.imglib2.outofbounds.OutOfBoundsMirrorFactory;
+import net.imglib2.outofbounds.OutOfBoundsMirrorFactory.Boundary;
+import net.imglib2.type.numeric.RealType;
+import net.imglib2.view.Views;
+
+/**
+ * Input is a 2- or 3-dimensional grayscales image.
+ * 
+ * Applies the Frangi Vesselness filter to an image to highlight vessel-like
+ * structures. The spacing parameter refers to the physical distance between
+ * data points in the RandomAccessibleInterval, and can vary for each dimension
+ * of the image.
+ *
+ * @author Gabe Selzer
+ */
+@Plugin(type = Ops.Filter.Vesselness.class)
+public class DefaultFrangi<T extends RealType<T>, U extends RealType<U>>
+		extends AbstractUnaryComputerOp<RandomAccessibleInterval<T>, RandomAccessibleInterval<U>>
+		implements Ops.Filter.Vesselness {
+
+	public final static int MIN_DIMS = 2;
+
+	public final static int MAX_DIMS = 3;
+
+	@Parameter
+	private double[] spacing;
+
+	@Parameter
+	private int scale;
+
+	protected double alpha = 0.5;
+	protected double beta = 0.5;
+
+	protected double minimumVesselness = Double.MIN_VALUE;
+	protected double maximumVesselness = Double.MAX_VALUE;
+
+	public double getMinimumVesselness() {
+		return minimumVesselness;
+	}
+
+	public double getMaximumVesselness() {
+		return maximumVesselness;
+	}
+
+	private double getDistance(RandomAccess<T> ahead, RandomAccess<T> behind, int d) {
+		double distance = 0;
+
+		for (int i = 0; i < d; i++) {
+			double separation = (ahead.getLongPosition(i) - behind.getLongPosition(i)) * spacing[i];
+			if (separation != 0) {
+				distance += (separation * separation);
+			}
+
+		}
+
+		return Math.sqrt(distance);
+	}
+
+	private double derivative(double val1, double val2, double distance) {
+		return ((val2 - val1) / distance);
+	}
+
+	public long numberOfPoints(RandomAccessibleInterval<T> image) {
```

Comment:
There is a [`Intervals.numElements(Dimensions)`](https://github.com/imglib/imglib2/blob/1941a7e90b70fb7e5149f8cb757cacb64ab182de/src/main/java/net/imglib2/util/Intervals.java#L519-L526) method in imglib2 that you could re-use.

Reply:
That is handy!

Reply:
Looking back, I am not actually sure why this is included in the class. I don't know if I meant to give this a purpose but I cannot find an instance in which this method is called. Should I just delete it?